primitives/src/persistent.rs

```diff
@@ -0,0 +1,240 @@
+//! Collections of wrappers around `rpds` crate to provide additional impls
+
+#[cfg(not(feature = "std"))]
+use alloc::{format, string::String};
+use core::{hash::Hash, mem};
+
+use derive_more::{Deref, DerefMut};
+use iroha_schema::{IntoSchema, TypeId};
+use parity_scale_codec::{Compact, Decode, Encode, Error, Input, Output};
+use serde::{Deserialize, Serialize};
+
+/// Persistent vector
+pub mod vector {
+    use super::*;
+
+    /// Persistent vector
+    #[derive(Serialize, Deserialize, Deref, DerefMut, Clone, Debug)]
+    pub struct PVector<T>(rpds::VectorSync<T>);
+
+    impl<T> PVector<T> {
+        /// Construct [`Self`]
+        pub fn new() -> Self {
+            Self(rpds::Vector::new_sync())
+        }
+    }
+
+    impl<T> Default for PVector<T> {
+        fn default() -> Self {
+            Self(rpds::VectorSync::default())
+        }
+    }
+
+    impl<T: Encode> Encode for PVector<T> {
+        #[allow(clippy::arithmetic_side_effects)]
+        fn size_hint(&self) -> usize {
+            mem::size_of::<u32>() + mem::size_of::<T>() * self.len()
+        }
+
+        fn encode_to<W: Output + ?Sized>(&self, dest: &mut W) {
+            let len = self.len();
+            Compact(u32::try_from(len).expect("Length is above u32::max")).encode_to(dest);
+
+            for element in self.iter() {
```

Comment:
I'd also consider:

```suggestion
            for element in self.iter()
                // Sanity check
                .take(len) {
```

Reply:
Imo it a bit too much since i think we can trust that iterator will return len elements.